src/v/raft/state_machine_manager.cc

```diff
@@ -0,0 +1,453 @@
+                auto snap = co_await serde::read_async<managed_snapshot>(
+                  parser);
+
+                for (auto& [key, buffer] : snap.snapshot_map) {
+                    auto it = _machines.find(key);
+                    if (
+                      it != _machines.end()
+                      && it->second->stm->last_applied_offset()
+                           < snapshot->metadata.last_included_index) {
+                        co_await it->second->stm->apply_raft_snapshot(buffer);
+                        it->second->stm->set_next(std::max(
+                          model::next_offset(last_offset),
+                          it->second->stm->next()));
+                    }
+                }
+            }
+            _next = model::next_offset(snapshot->metadata.last_included_index);
+        } catch (...) {
+            vlog(
+              _log.error,
+              "error applying raft snapshot - {}",
+              std::current_exception());
+        }
+        co_await snapshot->close();
+        if (e) {
+            // we thrown an exception here if any of the apply operations
+            // failed. Applying snapshot is idempotent and can be retried
+            std::rethrow_exception(e);
+        }
+    }
+}
+
+ss::future<> state_machine_manager::apply() {
+    try {
+        ss::coroutine::switch_to sg_sw(_apply_sg);
+        // wait until consensus commit index is >= _next
+        co_await _raft->events().wait(_next, model::no_timeout, _as);
+        auto u = co_await _apply_mutex.get_units();
+
+        if (_next < _raft->start_offset()) {
+            /**
+             * We need to return here as applied snapshot may not yet be
+             * committed.
+             */
+            co_return co_await apply_raft_snapshot();
+        }
+
+        /**
+         * Raft make_reader method allows callers reading up to
+         * last_visible index. In order to make the STMs safe and working
+         * with the raft semantics (i.e. what is applied must be comitted)
+         * we have to limit reading to the committed offset.
+         */
+        vlog(
+          _log.trace,
+          "reading batches in range [{}, {}]",
+          _next,
+          _raft->committed_offset());
+        /**
+         * Use default priority for now, it is going to be unified with apply
+         * scheduling group soon
+         */
+        storage::log_reader_config config(
+          _next, _raft->committed_offset(), ss::default_priority_class());
+
+        model::record_batch_reader reader = co_await _raft->make_reader(config);
+        // collect STMs which has the same _next offset as the offset in
+        // manager and there is no background apply taking place
+        std::vector<entry_ptr> machines;
+        for (auto& [_, entry] : _machines) {
+            if (
+              entry->stm->next() == _next
+              && entry->background_apply_mutex.ready()) {
```

Comment:
You are right, it could race with grabbing a mutex in background fiber as there was a scheduling point in between. I changed it so that when `maybe_start_background_apply` is called there are no scheduling points between the call and `mutex::get_units()`. `is_ready()` checks if there are no waiters on the semaphore so it should be fine